cachi2/core/package_managers/pip.py

```diff
@@ -1419,72 +1421,102 @@ def _download_dependencies(
     pip_deps_dir = output_dir.join_within_root("deps", "pip")
     pip_deps_dir.path.mkdir(parents=True, exist_ok=True)
 
-    downloaded: list[dict[str, Any]] = []
-    to_download: list[DistributionPackageInfo] = []
+    def _check_local(
+        artifacts: list[DistributionPackageInfo],
+    ) -> tuple[list[DistributionPackageInfo], list[DistributionPackageInfo]]:
+        # check if artifacts already exist locally
+        to_download: list[DistributionPackageInfo] = []
+        to_check: list[DistributionPackageInfo] = []
+        if artifacts:
+            for artifact in artifacts:
+                if artifact.path.exists():
+                    to_check.append(artifact)
+                    log.info("Artifact '%s' found locally", artifact.path.name)
+                    continue
+                else:
+                    to_download.append(artifact)
+        return to_download, to_check
+
+    def _hash_verify(
+        download_info: dict[str, Any], dpi: Optional[DistributionPackageInfo] = None
+    ) -> dict[str, Any]:
+        if dpi:
+            download_info = dpi.download_info
+
+            download_info["hash_verified"] = False
+            if dpi.should_verify_checksums():
+                try:
+                    must_match_any_checksum(dpi.path, dpi.checksums_to_verify)
+                    download_info["hash_verified"] = True
+                except PackageRejected:
+                    dpi.path.unlink()
+                    log.warning("Package '%s' is removed from the output directory", dpi.path.name)
+        else:
+            download_info["hash_verified"] = False
+            if require_hashes or req.kind == "url":
+                hashes = req.hashes or [req.qualifiers.get("cachito_hash", "")]
+                must_match_any_checksum(download_info["path"], list(map(_to_checksum_info, hashes)))
+                download_info["hash_verified"] = True
+
+        return download_info
+
+    def _finalize(req: PipRequirement, download_info: dict[str, Any]) -> dict[str, Any]:
+        log.info(
+            "Successfully downloaded/checked '%s' in path '%s'",
+            req.download_line,
+            download_info["path"].relative_to(output_dir),
+        )
+        download_info["kind"] = req.kind
+        download_info["requirement_file"] = str(requirements_file.file_path.subpath_from_root)
+        return download_info
```

Comment:
> it would be great to have such a helper defined outside of _download_dependencies

The refactor would explode in complexity because of the need to pass a huge amount of context *out of* `_download_dependencies`

> download_info ... would not need to be passed around

The pattern for all artifacts looks like

```python
download_info = get_download_info_from_external_source
download_info = munge_download_info  # hashes, kind, requirements file
downloaded.append(download_info)  # hopefully self-evident
```

so, *some* passing around is required